examples/Custom_Hooks/probability_of_improvement.py

```diff
@@ -0,0 +1,219 @@
+    train_y = objective.transform(measurements)
+    acqf = ProbabilityOfImprovement()
+    botorch_acqf = acqf.to_botorch(self.surrogate_model, searchspace, train_x, train_y)
+    comp_rep_tensor = to_tensor(searchspace.discrete.comp_rep).unsqueeze(1)
+    pi = botorch_acqf(comp_rep_tensor)
+    pi = pi + 1e-10 * torch.randn(pi.shape)
+    pi_list.append(pi)
+
+
+# Additionally, we define a function that plots the `poi_list` after all recommend  iterations:
+
+
+def plot_pi(
+    pi_list: list[torch.Tensor],
+    ax: Axes,
+    base_name: str,
+    path: Path = Path("."),
+) -> Figure:
+    """Plot the probability of improvement in 3D."""
+    cmap = plt.get_cmap("viridis")
+
+    pi_max = max([torch.max(p).item() for p in pi_list])
+
+    # Plot each PI tensor separately
+    for i, p in enumerate(pi_list):
+        pi_np = p.detach().numpy() if p.requires_grad else p.numpy()
+        x = np.linspace(0, pi_max, 500)
+        kde = gaussian_kde(pi_np)
+        y = kde(x)
+        z = np.full_like(y, i)
+
+        # Fill under the curve
+        verts = []
+        verts.append([(x[i], 0.0), *zip(x, y), (x[-1], 0.0)])
+        color = cmap(float(i) / len(pi_list))
+        poly = PolyCollection(verts, color=color, alpha=0.9)
+        ax.add_collection3d(poly, zs=i, zdir="x")
+
+        ax.plot(x, y, z, zdir="x", color=color)
+
+    # Set viewing angle
+    ax.view_init(elev=20, azim=30)
+
+    # Reduce space between the iterations
+    ax.set_box_aspect([0.7, 1, 1])
+
+    # Set the y-axis limit based on the maximal PI
+    ax.set_ylim(0, pi_max)
+
+    # Set x-axis ticks to have the correct iteration number
+    ax.set_xticks(np.arange(0, len(pi_list), 1))
+    ax.set_xticklabels([i for i in range(1, len(pi_list) + 1)])
+
+    ax.set_ylabel("PI")
+    ax.set_xlabel("Iteration")
+    ax.set_zlabel("Density")
+    ax.set_title("Investigate the PI")
+
+    output_path = Path(path, base_name)
+    # mypy thinks that ax.figure might become None, hence the explicit ignore
+    if isinstance(ax.figure, Figure):
+        ax.figure.savefig(
+            output_path,
+            format="svg",
+            transparent=True,
+        )
+    else:
+        warnings.warn("Plots could not be saved.")
+    plt.close()
+    return ax.get_figure()
+
+
+### Monkeypatching
+
+# Next, we create our recommender and monkeypatch its `recommend` method:
+
+my_recommender = TwoPhaseMetaRecommender(
+    initial_recommender=RandomRecommender(),
+    recommender=BotorchRecommender(
+        surrogate_model=GaussianProcessSurrogate(),
+        allow_repeated_recommendations=True,
+        allow_recommending_already_measured=True,
+    ),
+)
+my_recommender.recommender.recommend = MethodType(
+    register_hooks(
+        BotorchRecommender.recommend,
+        post_hooks=[check_probability_of_improvement],
+    ),
+    my_recommender.recommender,
+)
+
+# In this example we use `MethodType` to bind the `RandomRecommender.recommend`
```

Comment:
copied over from basic example, we dont patch the ranomd recommender, we patcht he botorch part

Reply:
I meant batch the recommend function of the random recommender, because we add our hook to it.

Reply:
Of couse it is the `BotorchRecommender`! Sorry for that!